src/frame/mod.rs

```diff
@@ -7,12 +7,90 @@ pub(crate) mod rtu;
 #[cfg(feature = "tcp")]
 pub(crate) mod tcp;
 
-use std::{borrow::Cow, error, fmt};
+use std::{
+    borrow::Cow,
+    error,
+    fmt::{self, Display},
+};
 
 use crate::bytes::Bytes;
 
-/// A Modbus function code is represented by an unsigned 8 bit integer.
-pub type FunctionCode = u8;
+/// A Modbus function code.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FunctionCode {
+    /// Modbus Function Code: `01` (`0x01`).
+    ReadCoils,
+    /// Modbus Function Code: `02` (`0x02`).
+    ReadDiscreteInputs,
+
+    /// Modbus Function Code: `05` (`0x05`).
+    WriteSingleCoil,
+    /// Modbus Function Code: `06` (`0x06`).
+    WriteSingleRegister,
+
+    /// Modbus Function Code: `03` (`0x03`).
+    ReadHoldingRegisters,
+    /// Modbus Function Code: `04` (`0x04`).
+    ReadInputRegisters,
+
+    /// Modbus Function Code: `15` (`0x0F`).
+    WriteMultipleCoils,
+    /// Modbus Function Code: `16` (`0x10`).
+    WriteMultipleRegisters,
+
+    /// Modbus Function Code: `22` (`0x16`).
+    MaskWriteRegister,
+
+    /// Modbus Function Code: `23` (`0x17`).
+    ReadWriteMultipleRegisters,
+
+    /// Custom Modbus Function Code.
+    Custom(u8),
+}
+
+impl FunctionCode {
+    /// Create a new [`FunctionCode`] with `value`.
+    #[must_use]
+    pub const fn new(value: u8) -> Self {
+        match value {
+            0x01 => FunctionCode::ReadCoils,
+            0x02 => FunctionCode::ReadDiscreteInputs,
+            0x05 => FunctionCode::WriteSingleCoil,
+            0x06 => FunctionCode::WriteSingleRegister,
+            0x03 => FunctionCode::ReadHoldingRegisters,
+            0x04 => FunctionCode::ReadInputRegisters,
+            0x0F => FunctionCode::WriteMultipleCoils,
+            0x10 => FunctionCode::WriteMultipleRegisters,
+            0x16 => FunctionCode::MaskWriteRegister,
+            0x17 => FunctionCode::ReadWriteMultipleRegisters,
+            code => FunctionCode::Custom(code),
+        }
+    }
+
+    /// Get the [`u8`] value of the current [`FunctionCode`].
+    #[must_use]
+    pub const fn value(self) -> u8 {
+        match self {
+            FunctionCode::ReadCoils => 0x01,
+            FunctionCode::ReadDiscreteInputs => 0x02,
+            FunctionCode::WriteSingleCoil => 0x05,
+            FunctionCode::WriteSingleRegister => 0x06,
+            FunctionCode::ReadHoldingRegisters => 0x03,
+            FunctionCode::ReadInputRegisters => 0x04,
+            FunctionCode::WriteMultipleCoils => 0x0F,
+            FunctionCode::WriteMultipleRegisters => 0x10,
+            FunctionCode::MaskWriteRegister => 0x16,
+            FunctionCode::ReadWriteMultipleRegisters => 0x17,
+            FunctionCode::Custom(code) => code,
+        }
+    }
+}
+
+impl Display for FunctionCode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.value())
```

Comment:
Directly and more idiomatic:

```suggestion
        self.value().fmt(f)
```